File: src/star.py
import math

def compute(args):
    center = args.center
    outerRadius = args.outerRadius
    innerRadius = args.innerRadius
    points = args.points
    
    numPoints = int(max(3, math.floor(points)))
    vertices = []
    angleStep = math.pi / numPoints
    outputNodes = {}

    for i in range(2 * numPoints):
        r = outerRadius if i % 2 == 0 else innerRadius
        angle = i * angleStep - math.pi / 2
        pt = {
            'type': 'Point',
            'x': center.x + r * math.cos(angle),
            'y': center.y + r * math.sin(angle)
        }
        outputNodes['p_{}'.format(i)] = pt
        vertices.append(pt)

    for i in range(2 * numPoints):
        outputNodes['l_{}'.format(i)] = {
            'type': 'Line',
            'p1': vertices[i],
            'p2': vertices[(i + 1) % (2 * numPoints)]
        }

    outputNodes['main'] = {
        'type': 'NStar',
        'center': center,
        'vertices': vertices
    }

    outputNodes['boundaryCircle'] = {
        'type': 'Circle',
        'center': center,
        'radius': { 'type': 'Scalar', 'value': outerRadius }
    }

    return outputNodes

NStarExt = {
    'name': 'N-Pointed Star',
    'keyword': 'n-star',
    'parameters': [
        { 'argName': 'center', 'type': 'Point', 'defaultValue': 'p0', 'variadic': False },
        { 'argName': 'outerRadius', 'type': 'Scalar', 'defaultValue': 4, 'variadic': False },
        { 'argName': 'innerRadius', 'type': 'Scalar', 'defaultValue': 2, 'variadic': False },
        { 'argName': 'points', 'type': 'Scalar', 'defaultValue': 5, 'variadic': False }
    ],
    'outputType': 'NStar',
    'compute': compute
}

# Export for ES6
__pragma__('js', 'export { NStarExt };')

File: src/star.ts
import { CircleNode, ExtensionDef, GeometricNode, LineNode, PointNode } from "./user-ext-def";

const OUTPUT_TYPE = 'NStar';
export const NStarExt: ExtensionDef<typeof OUTPUT_TYPE> = {
  name: 'N-Pointed Star',
  keyword: 'n-star',
  parameters: [
    { argName: 'center', type: 'Point', defaultValue: 'p0' },
    { argName: 'outerRadius', type: 'Scalar', defaultValue: 4 },
    { argName: 'innerRadius', type: 'Scalar', defaultValue: 2 },
    { argName: 'points', type: 'Scalar', defaultValue: 5 }
  ],
  outputType: OUTPUT_TYPE,

  compute: ({ center, outerRadius, innerRadius, points }) => {
    const numPoints = Math.max(3, Math.floor(points));
    const vertices = [];
    const angleStep = Math.PI / numPoints;
    const outputNodes: Record<string, GeometricNode> = {};

    for (let i = 0; i < 2 * numPoints; i++) {
      const r = i % 2 === 0 ? outerRadius : innerRadius;
      const angle = i * angleStep - Math.PI / 2;
      const pt: PointNode = {
        type: 'Point',
        x: center.x + r * Math.cos(angle),
        y: center.y + r * Math.sin(angle)
      };
      outputNodes[`p_${i}`] = pt;
      vertices.push(pt);
    }

    for (let i = 0; i < 2 * numPoints; i++) {
      outputNodes[`l_${i}`] = {
        type: 'Line',
        p1: vertices[i],
        p2: vertices[(i + 1) % (2 * numPoints)]
      } as LineNode;
    }

    outputNodes['main'] = {
      type: OUTPUT_TYPE,
      center,
      vertices
    };

    outputNodes['boundaryCircle'] = {
      type: 'Circle',
      center,
      radius: { type: 'Scalar', value: outerRadius }
    } as CircleNode;

    return outputNodes;
  }
};

